# Introduction to Pandas
Pandas is the default library for working with relational data.  The basic object in pandas is a dataframe, "a two-dimensional tabular, column-oriented data structure with both row and column labels."

In [1]:
import pandas as pd
import numpy as np
import random

# Loading Data Into Pandas 
Before doing anything else, we need to know how to load our data into memory.  The most common file in which relational data is stored is a csv.  Pandas includes a built in function to load data from a csv file into a dataframe.

In [2]:
df = pd.read_csv('boston_housing.csv')
print('dataframe shape: ', df.shape)
df.head()

dataframe shape:  (506, 14)


CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   PTRATIO       B  LSTAT  MEDV  
0     15.3  396.90   4.98  24.0  
1     17.8  396.90   9.14  21.6  
2     17.8  392.83   4.03  34.7  
3     18.7  394.63   2.94  33.4  
4     18.7  396.90   5.33  36.2

We can take a peek at the first 5 rows of the dataframe using the method DataFrame.head()

In [3]:
df.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   PTRATIO       B  LSTAT  MEDV  
0     15.3  396.90   4.98  24.0  
1     17.8  396.90   9.14  21.6  
2     17.8  392.83   4.03  34.7  
3     18.7  394.63   2.94  33.4  
4     18.7  396.90   5.33  36.2

We can verify that the number of rows and columns matches our expectations using the attribute DataFrame.shape

In [4]:
df.shape

(506, 14)

We can get the number of rows in our dataframe using the function len()

In [5]:
len(df)

506

## Common pitfalls loading data into pandas

### Specifying delimiters
Sometimes csvs use a character other than a comma as a separator, such as a semicolon or pipe.  This can lead to errors if we try to read the csv like normal.

In [6]:
delimdf = pd.read_csv('boston_housing_delimiter.csv')
print(delimdf.shape)
delimdf.head()

(506, 1)


CRIM;ZN;INDUS;CHAS;NOX;RM;AGE;DIS;RAD;TAX;PTRATIO;B;LSTAT;MEDV
0  0.00632;18.0;2.31;0.0;0.5379999999999999;6.575...            
1  0.02731;0.0;7.07;0.0;0.469;6.421;78.9;4.9671;2...            
2  0.02729;0.0;7.07;0.0;0.469;7.185;61.1;4.9671;2...            
3  0.03237;0.0;2.18;0.0;0.458;6.997999999999998;4...            
4  0.06905;0.0;2.18;0.0;0.458;7.147;54.2;6.0622;3...

<br>
Notice that we still have 506 rows but only one column.  If we look at the values we notice that they are separated by a semicolon.  The function pd.read_csv has an argument sep with a default value of ','.


In [7]:
delimdf = pd.read_csv('boston_housing_delimiter.csv', sep=';')
print(delimdf.shape)
delimdf.head()

(506, 14)


CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   PTRATIO       B  LSTAT  MEDV  
0     15.3  396.90   4.98  24.0  
1     17.8  396.90   9.14  21.6  
2     17.8  392.83   4.03  34.7  
3     18.7  394.63   2.94  33.4  
4     18.7  396.90   5.33  36.2

### Encoding errors
Most csvs you'll encounter use UTF-8 encoding, which is the default for the pd.read_csv() function.  However sometimes you'll get an encoding error.

In [8]:
encdf = pd.read_csv('boston_housing_encoding.csv', encoding='latin1')
print(encdf.shape)
encdf.head()

# this didn't give the error that I expected
# I'll either find a csv with a different encoding or skip the example for this section

(506, 14)


CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   PTRATIO       B  LSTAT  MEDV  
0     15.3  396.90   4.98  24.0  
1     17.8  396.90   9.14  21.6  
2     17.8  392.83   4.03  34.7  
3     18.7  394.63   2.94  33.4  
4     18.7  396.90   5.33  36.2

Encoding errors can be fixed by specifying the encoding argument in the pd.read_csv() function.  After 'utf-8', the next most common encoding in my experience is 'latin1', which is especially common in government documents.  After that good luck.

More info on encoding: https://www.joelonsoftware.com/2003/10/08/the-absolute-minimum-every-software-developer-absolutely-positively-must-know-about-unicode-and-character-sets-no-excuses/

## Loading data from a url

You can point pd.read_csv() to a url instead of a file on your computer.

## Other ways to read data

If you have data in another common file type there is likely a pandas function for reading it into a dataframe, for example pd.read_excel() and pd.read_json()

# Slicing Data in Pandas

Columns in dataframes are a data type called Series, which are similar to a numpy array.  To get a specific column from a dataframe, the syntax is dataframe[column_name] or dataframe.column_name

In [9]:
print(type(df['MEDV']))
df.MEDV.head()

<class 'pandas.core.series.Series'>


0    24.0
1    21.6
2    34.7
3    33.4
4    36.2
Name: MEDV, dtype: float64

Multiple columns can be selected using a list.

In [10]:
df[['AGE', 'DIS', 'RAD', 'TAX']].head()

AGE     DIS  RAD    TAX
0  65.2  4.0900  1.0  296.0
1  78.9  4.9671  2.0  242.0
2  61.1  4.9671  2.0  242.0
3  45.8  6.0622  3.0  222.0
4  54.2  6.0622  3.0  222.0

We pass a Boolean series to the dataframe to filter the dataframe to rows that meet a specific condition.  For example, first we get a series from the age column of proportions greater than 95.

In [11]:
df.AGE > 95

0      False
1      False
2      False
3      False
4      False
       ...  
501    False
502    False
503    False
504    False
505    False
Name: AGE, Length: 506, dtype: bool

We can pass this series into the dataframe like we did the list of columns.

In [12]:
df[df.AGE > 95].head()

CRIM    ZN  INDUS  CHAS    NOX     RM    AGE     DIS  RAD    TAX  \
7   0.14455  12.5   7.87   0.0  0.524  6.172   96.1  5.9505  5.0  311.0   
8   0.21124  12.5   7.87   0.0  0.524  5.631  100.0  6.0821  5.0  311.0   
20  1.25179   0.0   8.14   0.0  0.538  5.570   98.1  3.7979  4.0  307.0   
23  0.98843   0.0   8.14   0.0  0.538  5.813  100.0  4.0952  4.0  307.0   
31  1.35472   0.0   8.14   0.0  0.538  6.072  100.0  4.1750  4.0  307.0   

    PTRATIO       B  LSTAT  MEDV  
7      15.2  396.90  19.15  27.1  
8      15.2  386.63  29.93  16.5  
20     21.0  376.57  21.02  13.6  
23     21.0  394.54  19.88  14.5  
31     21.0  376.73  13.04  14.5

If we want only the MEDV of the rows that have an AGE > 95

In [13]:
df[df.AGE > 95].MEDV.head()

7     27.1
8     16.5
20    13.6
23    14.5
31    14.5
Name: MEDV, dtype: float64

You can compare to all sorts of things and combine them using & and |

In [14]:
df[(df.RAD == 5.0) & (df.DIS > df.DIS.mean())].head()

CRIM    ZN  INDUS  CHAS    NOX     RM    AGE     DIS  RAD    TAX  \
6   0.08829  12.5   7.87   0.0  0.524  6.012   66.6  5.5605  5.0  311.0   
7   0.14455  12.5   7.87   0.0  0.524  6.172   96.1  5.9505  5.0  311.0   
8   0.21124  12.5   7.87   0.0  0.524  5.631  100.0  6.0821  5.0  311.0   
9   0.17004  12.5   7.87   0.0  0.524  6.004   85.9  6.5921  5.0  311.0   
10  0.22489  12.5   7.87   0.0  0.524  6.377   94.3  6.3467  5.0  311.0   

    PTRATIO       B  LSTAT  MEDV  
6      15.2  395.60  12.43  22.9  
7      15.2  396.90  19.15  27.1  
8      15.2  386.63  29.93  16.5  
9      15.2  386.71  17.10  18.9  
10     15.2  392.52  20.45  15.0

# Useful Methods on Dataframes

DataFrame.describe() provides useful summary statistics about each of the columns in the dataframe, DataFrame.info() provides information about data types

In [15]:
df.describe()

CRIM          ZN       INDUS        CHAS         NOX          RM  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean     3.613524   11.363636   11.136779    0.069170    0.554695    6.284634   
std      8.601545   23.322453    6.860353    0.253994    0.115878    0.702617   
min      0.006320    0.000000    0.460000    0.000000    0.385000    3.561000   
25%      0.082045    0.000000    5.190000    0.000000    0.449000    5.885500   
50%      0.256510    0.000000    9.690000    0.000000    0.538000    6.208500   
75%      3.677083   12.500000   18.100000    0.000000    0.624000    6.623500   
max     88.976200  100.000000   27.740000    1.000000    0.871000    8.780000   

              AGE         DIS         RAD         TAX     PTRATIO           B  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean    68.574901    3.795043    9.549407  408.237154   18.455534  356.674032   
std     28.148861    2.105710    8.707259  168.537116    2.164946   91.294864   
min      2.900000    1.129600    1.000000  187.000000   12.600000    0.320000   
25%     45.025000    2.100175    4.000000  279.000000   17.400000  375.377500   
50%     77.500000    3.207450    5.000000  330.000000   19.050000  391.440000   
75%     94.075000    5.188425   24.000000  666.000000   20.200000  396.225000   
max    100.000000   12.126500   24.000000  711.000000   22.000000  396.900000   

            LSTAT        MEDV  
count  506.000000  506.000000  
mean    12.653063   22.532806  
std      7.141062    9.197104  
min      1.730000    5.000000  
25%      6.950000   17.025000  
50%     11.360000   21.200000  
75%     16.955000   25.000000  
max     37.970000   50.000000

In [16]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 14 columns):
CRIM       506 non-null float64
ZN         506 non-null float64
INDUS      506 non-null float64
CHAS       506 non-null float64
NOX        506 non-null float64
RM         506 non-null float64
AGE        506 non-null float64
DIS        506 non-null float64
RAD        506 non-null float64
TAX        506 non-null float64
PTRATIO    506 non-null float64
B          506 non-null float64
LSTAT      506 non-null float64
MEDV       506 non-null float64
dtypes: float64(14)
memory usage: 55.4 KB


It's often helpful to have a list of columns in your dataframe, for example to iterate over.

In [17]:
df.columns

Index(['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD', 'TAX',
       'PTRATIO', 'B', 'LSTAT', 'MEDV'],
      dtype='object')

In [18]:
for column_name in df.columns:
    print(column_name, df[column_name].sum())

CRIM 1828.4429200000004
ZN 5750.0
INDUS 5635.209999999999
CHAS 35.0
NOX 280.6757
RM 3180.025
AGE 34698.9
DIS 1920.2916
RAD 4832.0
TAX 206568.0
PTRATIO 9338.5
B 180477.06000000003
LSTAT 6402.450000000001
MEDV 11401.600000000002


Data often comes with column names that are either hard to understand or too long to consistently type out. Dataframe.rename() let's us rename the columns.  There are a couple ways to use it, but the easiest to me is to pass to the columns argument a dictionary with the old name as a key and the new name as a value.  This also lets us practice dictionary comprehension.

In [19]:
df.rename(columns = {key:key.lower() for key in df.columns}).head()

crim    zn  indus  chas    nox     rm   age     dis  rad    tax  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   ptratio       b  lstat  medv  
0     15.3  396.90   4.98  24.0  
1     17.8  396.90   9.14  21.6  
2     17.8  392.83   4.03  34.7  
3     18.7  394.63   2.94  33.4  
4     18.7  396.90   5.33  36.2

Methods like rename usually return a dataframe.  Above, the dataframe was returned, but our original dataframe df was not actually altered.  To make the change permanent, we can set df equal to the returned dataframe or we can use the argument inplace = True.

It's sometimes helpful to have a copy of a dataframe with DataFrame.copy()

When you make changes to a dataframe and want to save it to an external file, you can use DataFrame.to_csv(filepath)

# Useful Methods on Series

Most of the methods that do the actual heavy lifting are on Series rather than dataframes.  Descriptive statistics like sum, mean, median, max are all very straightforward.

In [15]:
df.MEDV.median()

21.2

Series.unique() returns an array with each unique value in the series.

In [21]:
df.RAD.unique()

array([ 1.,  2.,  3.,  5.,  4.,  8.,  6.,  7., 24.])

Series.value_counts() returns a series with the value as an index and the number of times that value occurs in the column as the value

In [98]:
df.RAD.value_counts()

24.0    132
5.0     115
4.0     110
3.0      38
6.0      26
8.0      24
2.0      24
1.0      20
7.0      17
Name: RAD, dtype: int64

Series.sort_values() returns a sorted series. Series.sort_values() also works on a dataframe, use the argument'by' to specify the column on which to sort the rest of the dataframe

In [105]:
df.MEDV.sort_values().head()

398    5.0
405    5.0
400    5.6
399    6.3
414    7.0
Name: MEDV, dtype: float64

In [106]:
df.sort_values(by = 'MEDV').head()

CRIM   ZN  INDUS  CHAS    NOX     RM    AGE     DIS   RAD    TAX  \
398  38.35180  0.0   18.1   0.0  0.693  5.453  100.0  1.4896  24.0  666.0   
405  67.92080  0.0   18.1   0.0  0.693  5.683  100.0  1.4254  24.0  666.0   
400  25.04610  0.0   18.1   0.0  0.693  5.987  100.0  1.5888  24.0  666.0   
399   9.91655  0.0   18.1   0.0  0.693  5.852   77.8  1.5004  24.0  666.0   
414  45.74610  0.0   18.1   0.0  0.693  4.519  100.0  1.6582  24.0  666.0   

     PTRATIO       B  LSTAT  MEDV  
398     20.2  396.90  30.59   5.0  
405     20.2  384.97  22.98   5.0  
400     20.2  396.90  26.77   5.6  
399     20.2  338.16  29.97   6.3  
414     20.2   88.27  36.98   7.0

series.duplicated() can be used to find duplicate values, which is useful for example when you want to check if a person's name appears twice in your data.

## Changing, Adding, and Dropping Columns

You can change a column in your dataframe simply using '='

When changing columns or adding new columns, pandas will yell at you if you use the df.column_name notation, so instead use df[column_name].

In [26]:
df['RAD'] = df.RAD.astype(int)
df.RAD.head()

0    1
1    2
2    2
3    3
4    3
Name: RAD, dtype: int64

In [27]:
df['RAD'] = df.RAD.astype(float)

You can add a new column to your dataframe simply by setting dataframe[new_column_name] = new_column

In [28]:
df['int_RAD'] = df.RAD.astype(int)

In [29]:
df.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   PTRATIO       B  LSTAT  MEDV  int_RAD  
0     15.3  396.90   4.98  24.0        1  
1     17.8  396.90   9.14  21.6        2  
2     17.8  392.83   4.03  34.7        2  
3     18.7  394.63   2.94  33.4        3  
4     18.7  396.90   5.33  36.2        3

You can delete an unwanted column by using df.drop(columns = [column_name1, column_name2])

In [30]:
df = df.drop(columns = 'int_RAD')

In [31]:
df.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   PTRATIO       B  LSTAT  MEDV  
0     15.3  396.90   4.98  24.0  
1     17.8  396.90   9.14  21.6  
2     17.8  392.83   4.03  34.7  
3     18.7  394.63   2.94  33.4  
4     18.7  396.90   5.33  36.2

## Apply

Some of the most powerful data transformations we can do make use of the apply method.  Apply is a very robust method that can do all sorts of things, but in a nutshell it takes a function and applies it to each element of a Series.  Apply can be used with functions you define or with lambda functions defined in place.

In [43]:
# creating a dummy variable for if the row has a MEDV greater than the mean MEDV value
df['MEDV_greater_than_mean'] = df.MEDV.apply(lambda x: 1 if x > df.MEDV.mean() else 0)
df.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   PTRATIO       B  LSTAT  MEDV  MEDV_greater_than_mean  
0     15.3  396.90   4.98  24.0                       1  
1     17.8  396.90   9.14  21.6                       0  
2     17.8  392.83   4.03  34.7                       1  
3     18.7  394.63   2.94  33.4                       1  
4     18.7  396.90   5.33  36.2                       1

In [45]:
df = df.drop(columns='MEDV_greater_than_mean')

You can call apply on a whole dataframe, using the axis argument to specify if you want the function to be applied to the rows or the columns.

## Groupby

Like apply, groupby is a robust method that can do a lot, roughly speaking it allows you to group a dataframe by the values in a column and apply a function to those groups.

In [41]:
# by itself groupby just returns a groupby object
df.groupby(by = 'RAD')

In [42]:
# but we can call methods on this groupby object like it's a dataframe
df.groupby(by='RAD').MEDV.mean()

RAD
1.0     24.365000
2.0     26.833333
3.0     27.928947
4.0     21.387273
5.0     25.706957
6.0     20.976923
7.0     27.105882
8.0     30.358333
24.0    16.403788
Name: MEDV, dtype: float64

## Dealing with Missing Values

Real world data sets often come with missing values and it's important to be able to deal with these.

First, let's introduce some missing values into our data.

In [69]:
zip_codes = [10001, 10002, 10003, 10004, 10005, np.nan]
df['zip_code'] = pd.Series([random.choice(zip_codes) for i in df.index])
df.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   PTRATIO       B  LSTAT  MEDV  zip_code  
0     15.3  396.90   4.98  24.0   10005.0  
1     17.8  396.90   9.14  21.6   10001.0  
2     17.8  392.83   4.03  34.7   10004.0  
3     18.7  394.63   2.94  33.4   10002.0  
4     18.7  396.90   5.33  36.2       NaN

If we call df.info() or df.describe() now, it will give us information about the missing numbers

In [70]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 15 columns):
CRIM        506 non-null float64
ZN          506 non-null float64
INDUS       506 non-null float64
CHAS        506 non-null float64
NOX         506 non-null float64
RM          506 non-null float64
AGE         506 non-null float64
DIS         506 non-null float64
RAD         506 non-null float64
TAX         506 non-null float64
PTRATIO     506 non-null float64
B           506 non-null float64
LSTAT       506 non-null float64
MEDV        506 non-null float64
zip_code    419 non-null float64
dtypes: float64(15)
memory usage: 59.4 KB


In [56]:
df.describe()

CRIM          ZN       INDUS        CHAS         NOX          RM  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean     3.613524   11.363636   11.136779    0.069170    0.554695    6.284634   
std      8.601545   23.322453    6.860353    0.253994    0.115878    0.702617   
min      0.006320    0.000000    0.460000    0.000000    0.385000    3.561000   
25%      0.082045    0.000000    5.190000    0.000000    0.449000    5.885500   
50%      0.256510    0.000000    9.690000    0.000000    0.538000    6.208500   
75%      3.677083   12.500000   18.100000    0.000000    0.624000    6.623500   
max     88.976200  100.000000   27.740000    1.000000    0.871000    8.780000   

              AGE         DIS         RAD         TAX     PTRATIO           B  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean    68.574901    3.795043    9.549407  408.237154   18.455534  356.674032   
std     28.148861    2.105710    8.707259  168.537116    2.164946   91.294864   
min      2.900000    1.129600    1.000000  187.000000   12.600000    0.320000   
25%     45.025000    2.100175    4.000000  279.000000   17.400000  375.377500   
50%     77.500000    3.207450    5.000000  330.000000   19.050000  391.440000   
75%     94.075000    5.188425   24.000000  666.000000   20.200000  396.225000   
max    100.000000   12.126500   24.000000  711.000000   22.000000  396.900000   

            LSTAT        MEDV  
count  506.000000  506.000000  
mean    12.653063   22.532806  
std      7.141062    9.197104  
min      1.730000    5.000000  
25%      6.950000   17.025000  
50%     11.360000   21.200000  
75%     16.955000   25.000000  
max     37.970000   50.000000

The method Series.isnull() returns a boolean series with True if the value is missing.

In [71]:
df.zip_code.isnull().head()

0    False
1    False
2    False
3    False
4     True
Name: zip_code, dtype: bool

We can count the number of null values in a column but just summing the boolean series.

In [72]:
df.zip_code.isnull().sum()

87

We can subset our dataframe to investigate the rows with missing values (in this case they are random but sometimes this might give you insight)

In [73]:
df[df.zip_code.isnull()].head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
4   0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   
5   0.02985   0.0   2.18   0.0  0.458  6.430  58.7  6.0622  3.0  222.0   
11  0.11747  12.5   7.87   0.0  0.524  6.009  82.9  6.2267  5.0  311.0   
24  0.75026   0.0   8.14   0.0  0.538  5.924  94.1  4.3996  4.0  307.0   
28  0.77299   0.0   8.14   0.0  0.538  6.495  94.4  4.4547  4.0  307.0   

    PTRATIO       B  LSTAT  MEDV  zip_code  
4      18.7  396.90   5.33  36.2       NaN  
5      18.7  394.12   5.21  28.7       NaN  
11     15.2  396.90  13.27  18.9       NaN  
24     21.0  394.33  16.30  15.6       NaN  
28     21.0  387.94  12.80  18.4       NaN

Many functions include arguments to specify what to do with nan values.

In [74]:
df.zip_code.value_counts()

10005.0    93
10003.0    87
10002.0    81
10004.0    79
10001.0    79
Name: zip_code, dtype: int64

In [75]:
df.zip_code.value_counts(dropna = False)

 10005.0    93
 10003.0    87
NaN         87
 10002.0    81
 10004.0    79
 10001.0    79
Name: zip_code, dtype: int64

Depending on the default value of this dropna argument, you may get errors.

If we don't want the rows with missing values in our dataframe, we can drop them entirely using DataFrame.dropna()

## Creating New Dataframes

Sometimes you need to create a dataframe from scratch.  My preferred way to do this is by using the pd.DataFrame.from_dict() function.  Let's say we look up information about population by zipcode. 

In [87]:
zip_dict = {'zip_code':[10001, 10002, 10003, 10004, 10005], 'population':[101200, 98700, 67500, 12900, 79400],
           'area':[10, 11, 12, 13, 14]}
zip_dict['pop_density'] = [round(i/j, 1) for i, j in zip(zip_dict['population'], zip_dict['area'])]
zip_df = pd.DataFrame.from_dict(zip_dict, orient = 'columns')
zip_df

zip_code  population  area  pop_density
0     10001      101200    10      10120.0
1     10002       98700    11       8972.7
2     10003       67500    12       5625.0
3     10004       12900    13        992.3
4     10005       79400    14       5671.4

## Merging Dataframes

While SQL is much more effecient for doing joins if all of your data is available in a database, sometimes you want to join data from two separate sources.  In this case, doing the join in python will be easier than trying to create a new database.  We can do data joins in python using the DataFrame.merge() function.  The important arguments are 'right' -- the data you want to join -- 'how' -- the type of join -- and 'on' -- the column on which to join.  The column names for on must be identical or must be specified using 'right_on' and 'left_on.'

In [88]:
df.merge(right = zip_df, how = 'inner', on='zip_code')

CRIM    ZN  INDUS  CHAS     NOX     RM    AGE     DIS   RAD    TAX  \
0     0.00632  18.0   2.31   0.0  0.5380  6.575   65.2  4.0900   1.0  296.0   
1     0.08829  12.5   7.87   0.0  0.5240  6.012   66.6  5.5605   5.0  311.0   
2     0.14455  12.5   7.87   0.0  0.5240  6.172   96.1  5.9505   5.0  311.0   
3     0.22489  12.5   7.87   0.0  0.5240  6.377   94.3  6.3467   5.0  311.0   
4     0.80271   0.0   8.14   0.0  0.5380  5.456   36.6  3.7965   4.0  307.0   
5     1.23247   0.0   8.14   0.0  0.5380  6.142   91.7  3.9769   4.0  307.0   
6     0.67191   0.0   8.14   0.0  0.5380  5.813   90.3  4.6820   4.0  307.0   
7     1.00245   0.0   8.14   0.0  0.5380  6.674   87.3  4.2390   4.0  307.0   
8     0.17142   0.0   6.91   0.0  0.4480  5.682   33.8  5.1004   3.0  233.0   
9     0.22927   0.0   6.91   0.0  0.4480  6.030   85.5  5.6894   3.0  233.0   
10    0.25387   0.0   6.91   0.0  0.4480  5.399   95.3  5.8700   3.0  233.0   
11    0.08873  21.0   5.64   0.0  0.4390  5.963   45.7  6.8147   4.0  243.0   
12    0.14932  25.0   5.13   0.0  0.4530  5.741   66.2  7.2254   8.0  284.0   
13    0.17171  25.0   5.13   0.0  0.4530  5.966   93.4  6.8185   8.0  284.0   
14    0.11027  25.0   5.13   0.0  0.4530  6.456   67.8  7.2255   8.0  284.0   
15    0.01951  17.5   1.38   0.0  0.4161  7.104   59.5  9.2229   3.0  216.0   
16    0.04379  80.0   3.37   0.0  0.3980  5.787   31.1  6.6115   4.0  337.0   
17    0.08707   0.0  12.83   0.0  0.4370  6.140   45.8  4.0905   5.0  398.0   
18    0.05059   0.0   4.49   0.0  0.4490  6.389   48.0  4.7794   3.0  247.0   
19    0.02875  28.0  15.04   0.0  0.4640  6.211   28.9  3.6659   4.0  270.0   
20    0.11504   0.0   2.89   0.0  0.4450  6.163   69.6  3.4952   2.0  276.0   
21    0.12083   0.0   2.89   0.0  0.4450  8.069   76.0  3.4952   2.0  276.0   
22    0.21161   0.0   8.56   0.0  0.5200  6.137   87.4  2.7147   5.0  384.0   
23    0.17120   0.0   8.56   0.0  0.5200  5.836   91.9  2.2110   5.0  384.0   
24    0.12802   0.0   8.56   0.0  0.5200  6.474   97.1  2.4329   5.0  384.0   
25    0.10084   0.0  10.01   0.0  0.5470  6.715   81.6  2.6775   6.0  432.0   
26    0.12329   0.0  10.01   0.0  0.5470  5.913   92.9  2.3534   6.0  432.0   
27    0.17134   0.0  10.01   0.0  0.5470  5.928   88.2  2.4631   6.0  432.0   
28    0.14476   0.0  10.01   0.0  0.5470  5.731   65.2  2.7592   6.0  432.0   
29    0.15038   0.0  25.65   0.0  0.5810  5.856   97.0  1.9444   2.0  188.0   
..        ...   ...    ...   ...     ...    ...    ...     ...   ...    ...   
389   9.23230   0.0  18.10   0.0  0.6310  6.216  100.0  1.1691  24.0  666.0   
390   8.26725   0.0  18.10   1.0  0.6680  5.875   89.6  1.1296  24.0  666.0   
391  18.49820   0.0  18.10   0.0  0.6680  4.138  100.0  1.1370  24.0  666.0   
392  14.33370   0.0  18.10   0.0  0.7000  4.880  100.0  1.5895  24.0  666.0   
393   6.96215   0.0  18.10   0.0  0.7000  5.713   97.0  1.9265  24.0  666.0   
394   8.64476   0.0  18.10   0.0  0.6930  6.193   92.6  1.7912  24.0  666.0   
395   5.87205   0.0  18.10   0.0  0.6930  6.405   96.0  1.6768  24.0  666.0   
396   9.91655   0.0  18.10   0.0  0.6930  5.852   77.8  1.5004  24.0  666.0   
397  14.23620   0.0  18.10   0.0  0.6930  6.343  100.0  1.5741  24.0  666.0   
398  24.80170   0.0  18.10   0.0  0.6930  5.349   96.0  1.7028  24.0  666.0   
399  18.08460   0.0  18.10   0.0  0.6790  6.434  100.0  1.8347  24.0  666.0   
400  37.66190   0.0  18.10   0.0  0.6790  6.202   78.7  1.8629  24.0  666.0   
401   7.36711   0.0  18.10   0.0  0.6790  6.193   78.1  1.9356  24.0  666.0   
402   9.33889   0.0  18.10   0.0  0.6790  6.380   95.6  1.9682  24.0  666.0   
403  13.91340   0.0  18.10   0.0  0.7130  6.208   95.0  2.2222  24.0  666.0   
404   7.52601   0.0  18.10   0.0  0.7130  6.417   98.3  2.1850  24.0  666.0   
405   6.71772   0.0  18.10   0.0  0.7130  6.749   92.6  2.3236  24.0  666.0   
406   4.66883   0.0  18.10   0.0  0.7130  5.976   87.9  2.5806  24.0  666.0   
407   8.20058   0.0  18.10   0.0  0.7130  5.936   80.3  2.7

In [91]:
df = df.merge(right = zip_df, how = 'left', on='zip_code')
df.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

       ...             B  LSTAT  MEDV  zip_code  population_x  area_x  \
0      ...        396.90   4.98  24.0   10005.0       79400.0    14.0   
1      ...        396.90   9.14  21.6   10001.0      101200.0    10.0   
2      ...        392.83   4.03  34.7   10004.0       12900.0    13.0   
3      ...        394.63   2.94  33.4   10002.0       98700.0    11.0   
4      ...        396.90   5.33  36.2       NaN           NaN     NaN   

   pop_density_x  population_y  area_y  pop_density_y  
0         5671.4       79400.0    14.0         5671.4  
1        10120.0      101200.0    10.0        10120.0  
2          992.3       12900.0    13.0          992.3  
3         8972.7       98700.0    11.0         8972.7  
4            NaN           NaN     NaN            NaN  

[5 rows x 21 columns]

In [93]:
# I made a mistake here thinking that merge was not inplace, I'll leave it since it's illustrative.